### PR TITLE
Right-size byte arrays in `InputStreamConsumingPublisher` to avoid excessive heap usage on small-chunk streams

### PR DESCRIPTION
**Pull request by alextwoods**
Right-size byte arrays in `InputStreamConsumingPublisher` to avoid unbounded heap growth on small-chunk streams

## Motivation and Context
`AsyncRequestBody.fromInputStream()` (used by S3 multipart uploads via `S3TransferManager` and `S3AsyncClient.multipartEnabled(true)`) feeds an `InputStream` through `InputStreamConsumingPublisher`, which allocates a fixed 16 KB `byte[]` for every `read()` call.

For sources that return fewer bytes per `read()` than the requested length — most notably `PipedInputStream` (default 1 KB pipe buffer), but also sockets and any wrapping `BufferedInputStream` that short-circuits on `available()` — the publisher emits a `ByteBuffer` whose backing array is still 16 KB even though only ~1 KB of it is useful data. Those oversized backing arrays then flow through:

* `RetryableSubAsyncRequestBody.buffers` (held until the part's S3 upload response arrives, for retry support)
* `SimplePublisher` queues
* Netty's `ChannelOutboundBuffer` via `Unpooled.wrappedBuffer`

resulting in a ~16x memory amplification: a 100 MB upload can allocate 1.6 GB of `byte[]` arrays, of which 1.5 GB is unused capacity that nonetheless pins memory until the upload completes.

## Modifications
`InputStreamConsumingPublisher`:

1. Use `InputStream.available()` as a sizing hint for the per-read allocation, clamped between a 1 KB minimum and the existing 16 KB maximum. Streams that report useful availability (e.g., `PipedInputStream`, sockets) get right-sized allocations upfront — no copy needed.
2. As a fallback for streams whose `available()` underestimates (or returns 0 because the pipe is momentarily empty), copy the data into a right-sized array if the read returned less than half the buffer size. This catches the worst-case waste without paying a copy cost on the common full-read path.
3. Both `available()` failures (`IOException` thrown, negative value returned) fall back to the original 16 KB allocation, preserving previous behavior on edge cases.

The fix is contained to `InputStreamConsumingPublisher` and does not change the publisher's reactive-streams contract (item ordering, demand handling, cancellation) — only the size of the backing array each emitted `ByteBuffer` references.

## Testing
Added new unit tests to cover correctness and edge cases as well as memory usage. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: utils/src/main/java/software/amazon/awssdk/utils/async/InputStreamConsumingPublisher.java [hunk 1]
```diff
@@ -73,6 +87,39 @@ public long doBlockingWrite(InputStream inputStream) {
...
+     * always report 0 (or that throw) fall back to {@link #BUFFER_SIZE}.
+     */
+    private static int allocSize(InputStream inputStream) {
+        try {
+            int avail = inputStream.available();
+            if (avail <= 0) {
+                return BUFFER_SIZE;
+            }
+            return Math.min(BUFFER_SIZE, Math.max(MIN_BUFFER_SIZE, avail));
+        } catch (IOException ignored) {
```
Comment (RanVaknin) [inline utils/src/main/java/software/amazon/awssdk/utils/async/InputStreamConsumingPublisher.java]: Can we change this to generic `} catch (Exception ignored) {` if someone has their own streaming implementation that throws something custom this will fail.

Comment (alextwoods) [reply utils/src/main/java/software/amazon/awssdk/utils/async/InputStreamConsumingPublisher.java]: Yeah, I think that makes sense since this is a net-new call.

**Review by RanVaknin — CHANGES_REQUESTED** [review]
PR looks good, the one thing that concerns me is using `.available()` as clue for resizing. In theory some downstream stream implementation might not return what we are expecting. I don't know if there's a way around using it, but worth calling out.

**Comment by alextwoods** [comment]
Using `available` is just a hint - when the downstream does behave well it helps optimize the memory usage, but the real backstop on excess memory usage is trimming the buffer before sending it.

**Review by RanVaknin — APPROVED** [review]
